### PR TITLE
add IBM VPC service instance client

### PR DESCRIPTION
**Pull request by omerb01**
add IBM VPC service instance client for start/stop interface using the docker executor.
its authentication needs to use IAM api key

example config usage:
```
pywren:
    storage_bucket: ***
  
ibm:
    iam_api_key: ***

docker:
    host: ***
    ssh_user: ***
    ssh_password: ***
    ssh_key_filename: < path > (OPTIONAL)
    remote_client: ibm_vpc

ibm_vpc:
    endpoint: ***
    instance_id: ***
    version: dd-mm-yyyy (OPTIONAL)
    generation: 1/2 (OPTIONAL)

ibm_cos:
    endpoint: ***
    access_key: ***
    secret_key: ***
```

example code:
```python
import pywren_ibm_cloud as pywren

def my_func(args):
    return 'Hello World'

pw = pywren.docker_executor(storage_backend='ibm_cos')
pw.call_async(my_func, '')
print(pw.get_result())
```

this code will start a VM instance provided by IBM VPC service and will invoke the hello world function into it. afterwards, it will stop the VM and will make sure that it is stopping.
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: pywren_ibm_cloud/compute/backends/docker/docker.py [hunk 1]
```diff
@@ -32,15 +33,31 @@ def __init__(self, docker_config):
...
         if self._is_localhost:
             try:
                 self.docker_client = docker.from_env()
             except Exception:
                 pass
 
+        if self._has_remote_client:
+
+            if self.config['remote_client'] == 'gen2':
+                remote_client_backend = 'IBMVPCInstanceClient'
```
Comment (gilv) [inline pywren_ibm_cloud/compute/backends/docker/docker.py:49]: @omerb01 this should be part of gen2 configuration as well

File: pywren_ibm_cloud/libs/docker/clients.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import os
```
Comment (gilv) [inline pywren_ibm_cloud/libs/docker/clients.py:1]: @omerb01 it should be one client per VM type... not all clients inside this one.

**Comment by omerb01** [comment]
@gilv thanks for the review, it is ready for merge

File: pywren_ibm_cloud/libs/docker/clients/ibm_vpc/ibm_vpc.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        if user_agent:
+            default_user_agent = self.session.headers['User-Agent']
+            headers['User-Agent'] = default_user_agent + ' {}'.format(user_agent)
+        self.session.headers.update(headers)
+
+        adapter = requests.adapters.HTTPAdapter()
+        self.session.mount('https://', adapter)
+
+    def _authorize_session(self):
+        self.config['token'], self.config['token_expiry_time'] = self.ibm_iam_api_key_manager.get_token()
```
Comment (gilv) [inline pywren_ibm_cloud/libs/docker/clients/ibm_vpc/ibm_vpc.py:33]: @omerb01 is this generate new token every time this method is called? I think it is... can you check it? you don't need Gen2 for this

Comment (omerb01) [reply pywren_ibm_cloud/libs/docker/clients/ibm_vpc/ibm_vpc.py:33]: @gilv nope, it will generate a new token only if it is expired :)

**Comment by gilv** [comment]
@omerb01 please update documentation also how to use and enable gen2

File: pywren_ibm_cloud/libs/docker/clients/ibm_vpc/ibm_vpc.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        self.session.headers['Authorization'] = 'Bearer ' + self.config['token']
+
+    def get_instance(self):
+        url = '/'.join([self.config['endpoint'], 'v1', 'instances', self.config['instance_id']
+                        + f'?version={self.config["version"]}&generation={self.config["generation"]}'])
+        self._authorize_session()
+        res = self.session.get(url)
```
Comment (gilv) [inline pywren_ibm_cloud/libs/docker/clients/ibm_vpc/ibm_vpc.py:40]: @omerb01 few lines above you update session headers with the token. If authorize session will create new token, the headers will not be updated.

Comment (omerb01) [reply pywren_ibm_cloud/libs/docker/clients/ibm_vpc/ibm_vpc.py:40]: @gilv I'll check that when we will have access for the service

File: pywren_ibm_cloud/compute/backends/docker/config.py [hunk 5]
```diff
@@ -34,9 +36,18 @@ def load_config(config_data):
         config_data['docker'] = {'host': 'localhost'}
 
     if config_data['docker']['host'] not in ['127.0.0.1', 'localhost']:
-        if 'ssh_user' not in config_data['docker'] or 'ssh_password' not in config_data['docker']:
+
+        required_credentials_0 = ('ssh_user', 'ssh_password')
+        required_credentials_1 = ('ssh_user', 'ssh_key_filename', 'ssh_passphrase')
```
Comment (gilv) [inline pywren_ibm_cloud/compute/backends/docker/config.py:41]: @omerb01 this suppose to be optional. Not everyone need to provide ssh_key_filename and ssh_passphrase

**Comment by gilv** [comment]
@omerb01 please create new doc file under https://github.com/pywren/pywren-ibm-cloud/tree/master/docs and add usage + example there (the same text you have in the desc of the pull request )

File: pywren_ibm_cloud/compute/backends/docker/config.py (hunk 5 above)
Comment (omerb01) [reply pywren_ibm_cloud/compute/backends/docker/config.py:41]: @gilv I disabled `ssh_passphrase` and made `ssh_key_filename` optional

**Comment by omerb01** [comment]
@gilv Done

File: pywren_ibm_cloud/compute/backends/docker/docker.py [hunk 6]
```diff
@@ -17,9 +18,24 @@
...
 
 
+class RemoteClientStartStopContextManager:
+    def __init__(self, remote_client=None):
+        self.remote_client = remote_client
+
+    def __enter__(self):
+        if self.remote_client:
+            self.remote_client.create_instance_action('start')
+            time.sleep(5)
```
Comment (gilv) [inline pywren_ibm_cloud/compute/backends/docker/docker.py:32]: @omerb01 the last one...please make it confirable as well. We can't use 5 as static in the code

**Comment by gilv** [comment]
@omerb01 great! thanks

**Pull request merged**